src/Neo/Ledger/Blockchain.cs

```diff
@@ -414,26 +416,47 @@ private void Persist(Block block)
                     all_application_executed.Add(application_executed);
                     transactionStates = engine.GetState<TransactionState[]>();
                 }
-                DataCache clonedSnapshot = snapshot.CreateSnapshot();
-                // Warning: Do not write into variable snapshot directly. Write into variable clonedSnapshot and commit instead.
+
+                var tasks = new List<Task<(TransactionState, DataCache, ApplicationExecuted)>>();
+
+                // Warning: Do not write into variable snapshot directly.
+                // Write into variable clonedSnapshot and commit instead.
                 foreach (TransactionState transactionState in transactionStates)
                 {
-                    Transaction tx = transactionState.Transaction;
-                    using ApplicationEngine engine = ApplicationEngine.Create(TriggerType.Application, tx, clonedSnapshot, block, system.Settings, tx.SystemFee);
-                    engine.LoadScript(tx.Script);
-                    transactionState.State = engine.Execute();
-                    if (transactionState.State == VMState.HALT)
+                    DataCache txSnapshot = snapshot.CreateSnapshot();
+                    var task = OnExecuteTransactionAsync(system, block, txSnapshot, transactionState);
+                    tasks.Add(task);
+                }
+
+                // var readSet = new HashSet<StorageKey>();
+                var writeSet = new HashSet<StorageKey>();
+
+                var randomUsed = false;
+                foreach (var task in tasks)
+                {
+                    var (transactionState, txSnapshot, executed) = task.Result;
+
+                    if (txSnapshot.GetReadSet().Overlaps(writeSet) || (randomUsed && txSnapshot.isRandomNumberCalled))
```

Comment:
What if I have tx A reading nothing and writing into K some V1 and tx B reading nothing and writing into K some V2?

Reply:
then tx B will reexecute

Reply:
Will, this is a feature anyway, for potential ecosystem boost~. I know this is useful only if we push to the neo limit. In reality, of not much use.